lib/byron/bench/Restore.hs

```diff
@@ -0,0 +1,460 @@
+                            tr
+                            socketPath
+                            "mainnet-seq.dat"
+                            (walletRnd @'Mainnet))
+
+                    , bench ("restore " <> "mainnet" <> " 1% ownership")
+                        (bench_restoration @'Mainnet @IcarusKey
+                            tr
+                            socketPath
+                            "mainnet-1-percent.dat"
+                            (initAnyState "Benchmark 1% Wallet" 0.01))
+
+                    , bench ("restore " <> "mainnet" <> " 2% ownership")
+                        (bench_restoration @'Mainnet @IcarusKey
+                            tr
+                            socketPath
+                            "mainnet-2-percent.dat"
+                            (initAnyState "Benchmark 2% Wallet" 0.02))
+                    ]
+  where
+    walletRnd
+        :: (WalletId, WalletName, RndState n)
+    walletRnd =
+        let
+            seed = SomeMnemonic . unsafeMkMnemonic @15 $ T.words
+                "involve key curtain arrest fortune custom lens marine before \
+                \material wheel glide cause weapon wrap"
+            xprv = Byron.generateKeyFromSeed seed mempty
+            wid = WalletId $ digest $ publicKey xprv
+            wname = WalletName "Benchmark Sequential Wallet"
+            rngSeed = 0
+            s = mkRndState xprv rngSeed
+        in
+            (wid, wname, s)
+
+runBenchmarks :: [IO (Text, Double)] -> IO ()
+runBenchmarks bs = do
+    initializeTime
+    -- NOTE: Adding an artificial delay between successive runs to get a better
+    -- output for the heap profiling.
+    rs <- forM bs $ \io -> io <* let _2s = 2000000 in threadDelay _2s
+    sayErr "\n\nAll results:"
+    mapM_ (uncurry printResult) rs
+
+bench :: Text -> IO () -> IO (Text, Double)
+bench benchName action = do
+    sayErr $ "Running " <> benchName
+    start <- getTime
+    res <- action
+    evaluate (rnf res)
+    finish <- getTime
+    let dur = finish - start
+    printResult benchName dur
+    pure (benchName, dur)
+
+printResult :: Text -> Double -> IO ()
+printResult benchName dur = sayErr . fmt $ "  "+|benchName|+": "+|secs dur|+""
+
+initBenchmarkLogging :: Severity -> IO (CM.Configuration, Trace IO Text)
+initBenchmarkLogging minSeverity = do
+    c <- defaultConfigStdout
+    CM.setMinSeverity c minSeverity
+    CM.setSetupBackends c [CM.KatipBK, CM.AggregationBK]
+    (tr, _sb) <- setupTrace_ c "bench-restore"
+    pure (c, tr)
+
+{-------------------------------------------------------------------------------
+                                  Benchmarks
+-------------------------------------------------------------------------------}
+
+{-# ANN bench_restoration ("HLint: ignore Use camelCase" :: String) #-}
+bench_restoration
+    :: forall (n :: NetworkDiscriminant) (k :: Depth -> * -> *) s t.
+        ( IsOurs s Address
+        , IsOurs s ChimericAccount
```

Comment:
Why is this constraint needed ?

Reply:
For [`createWallet`](https://github.com/input-output-hk/cardano-wallet/blob/a3fecaccdb0600b0a5c76b1027441581cd2f43e6/lib/core/src/Cardano/Wallet.hs#L482), all the way to [`prefilterBlock`](https://github.com/input-output-hk/cardano-wallet/blob/a3fecaccdb0600b0a5c76b1027441581cd2f43e6/lib/core/src/Cardano/Wallet/Primitive/Model.hs#L370-L377).